astropy/coordinates/baseframe.py

```diff
@@ -424,39 +424,121 @@ def __init__(self, *args, copy=True, representation_type=None,
-                # TODO: assumes the differential class is for the velocity
-                # differential
-                differential_type = {'s': differential_type}
+        # Set frame attributes, if any
 
-            elif isinstance(differential_type, str):
-                # TODO: assumes the differential class is for the velocity
-                # differential
-                diff_cls = r.DIFFERENTIAL_CLASSES[differential_type]
-                differential_type = {'s': diff_cls}
+        values = {}
+        for fnm, fdefault in self.get_frame_attr_names().items():
+            # Read-only frame attributes are defined as FrameAttribute
+            # descriptors which are not settable, so set 'real' attributes as
+            # the name prefaced with an underscore.
+
+            if fnm in kwargs:
+                value = kwargs.pop(fnm)
+                setattr(self, '_' + fnm, value)
+                # Validate attribute by getting it. If the instance has data,
+                # this also checks its shape is OK. If not, we do it below.
+                values[fnm] = getattr(self, fnm)
+            else:
+                setattr(self, '_' + fnm, fdefault)
+                self._attr_names_with_defaults.append(fnm)
+
+        if kwargs:
+            raise TypeError(
+                f'Coordinate frame got unexpected keywords: {list(kwargs)}')
+
+        # We do ``is None`` because self._data might evaluate to false for
+        # empty arrays or data == 0
+        if self._data is None:
+            # No data: we still need to check that any non-scalar attributes
+            # have consistent shapes. Collect them for all attributes with
+            # size > 1 (which should be array-like and thus have a shape).
+            shapes = {fnm: value.shape for fnm, value in values.items()
+                      if getattr(value, 'shape', ())}
+            if shapes:
+                if len(shapes) > 1:
+                    try:
+                        self._no_data_shape = check_broadcast(*shapes.values())
+                    except ValueError:
+                        raise ValueError(
+                            f"non-scalar attributes with inconsistent shapes: {shapes}")
 
-            elif differential_type is None:
-                if representation_type == self.default_representation:
-                    differential_type = {'s': self.default_differential}
+                    # Above, we checked that it is possible to broadcast all
+                    # shapes.  By getting and thus validating the attributes,
+                    # we verify that the attributes can in fact be broadcast.
+                    for fnm in shapes:
+                        getattr(self, fnm)
                 else:
-                    differential_type = {'s': 'base'}  # see set_representation_cls()
+                    self._no_data_shape = shapes.popitem()[1]
+
+            else:
+                self._no_data_shape = ()
+
+        # else:
```

Comment:
I sort-of see your point of not using `else` - though in a way both clauses are "wrap-up" for the two cases. In any case, don't leave it commented out here!

Aside: I do wonder whether this is actually needed - but a good thorough look at the cache is also for another time!

Reply:
> In any case, don't leave it commented out here!

Fair enough, removing it!

> Aside: I do wonder whether this is actually needed - but a good thorough look at the cache is also for another time!

Definitely - I still do not understand what that chunk does :sweat_smile: